src/typechecker.erl

```diff
@@ -3438,16 +3438,76 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_call(erlang:anno(), atom(), list()) -> map().
+check_guard_call(P, is_atom, [{var, _, Var}]) -> #{Var => {type, P, atom, []}};
+check_guard_call(P, is_binary, [{var, _, Var}]) -> #{Var => {type, P, binary, []}};
```

Comment:
These are using the annotation of the function call for the created type, which we don't usually do. There are helper functions `type/1,2` to create types with an empty annotation, e.g. `type(atom)`.